File: src/scripts/lint_role_experiences.py
#!/usr/bin/env python3
"""Linter for ``agents/roles/<slug>/`` role experiences.

Asserts the structural floor pinned in
``docs/contracts/role-experience.md``:

  1. Every role directory has ``index.md`` + ``skills.yml``.
  2. ``index.md`` carries all required frontmatter keys (``role``,
     ``display_name``, ``tagline``, ``recommended_packs``,
     ``install_path_hint``, ``recruit_session_ref``, ``status``).
  3. ``index.md`` body declares at least three first tasks (per
     ``## Three first tasks`` heading + numbered list).
  4. ``prompts/`` contains at least five prompts (per the roadmap's
     Phase B Step 6 acceptance — ``≥ 5 prompts per role``).
  5. Every prompt under ``prompts/<name>.md`` carries the four
     required frontmatter keys (``name``, ``intent``, ``inputs``,
     ``output_shape``) plus ``skill_hint``.
  6. Every ``skills.yml`` skill ``id`` resolves to an existing skill
     under ``.agent-src.uncondensed/skills/<id>/`` or
     ``dist/agent-src/skills/<id>/`` (condensed projection counts).

The roadmap step is intentionally opt-in for CI — the existing
``task ci`` wiring is the caller's concern; this script is the
mechanical floor it can wrap.

Phase B Step 6 of ``road-to-frictionless-employee-workspace.md``.
Adds a ``--plain-language`` mode for Phase C Step 5 that scans the
role indices for jargon listed in
``docs/contracts/plain-language-surface.md`` and flags any hit.
"""
from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

import yaml

ROOT = Path(__file__).resolve().parents[2]
ROLES_DIR = ROOT / "agents" / "roles"
SKILL_SOURCES = [
    ROOT / ".agent-src.uncondensed" / "skills",
    ROOT / "dist/agent-src" / "skills",
]

REQUIRED_INDEX_KEYS = {
    "role",
    "display_name",
    "tagline",
    "recommended_packs",
    "install_path_hint",
    "recruit_session_ref",
    "status",
}

REQUIRED_PROMPT_KEYS = {"name", "intent", "inputs", "output_shape", "skill_hint"}

MIN_FIRST_TASKS = 3
MIN_PROMPTS_PER_ROLE = 5

FRONTMATTER_RE = re.compile(r"^---\r?\n(.*?)\r?\n---\r?\n", re.DOTALL)
TASK_HEADING_RE = re.compile(r"^## Three first tasks\s*$", re.MULTILINE)
TASK_ITEM_RE = re.compile(r"^\s*\d+\.\s+\*\*[^*]+\*\*", re.MULTILINE)

# Jargon list mirrors docs/contracts/plain-language-surface.md.
PLAIN_LANGUAGE_JARGON = [
    "council",
    "trust level",
    "pack",
    "orchestration",
    "contract",
    "advisory",
]


def parse_frontmatter(text: str) -> tuple[dict, str]:
    m = FRONTMATTER_RE.match(text)
    if m is None:
        return {}, text
    try:
        loaded = yaml.safe_load(m.group(1)) or {}
    except yaml.YAMLError:
        return {}, text[m.end():]
    if not isinstance(loaded, dict):
        return {}, text[m.end():]
    return loaded, text[m.end():]


def count_first_tasks(body: str) -> int:
    heading_match = TASK_HEADING_RE.search(body)
    if heading_match is None:
        return 0
    # Slice from heading to the next "## " heading (or end of file).
    after = body[heading_match.end():]
    next_section = re.search(r"^## ", after, re.MULTILINE)
    section = after[: next_section.start()] if next_section else after
    return len(TASK_ITEM_RE.findall(section))


def all_skills() -> set[str]:
    found: set[str] = set()
    for src in SKILL_SOURCES:
        if not src.exists():
            continue
        for child in src.iterdir():
            if child.is_dir():
                found.add(child.name)
    return found


def lint_role(role_dir: Path, known_skills: set[str], failures: list[str]) -> None:
    slug = role_dir.name
    if slug.startswith(("_", ".")):
        return

    index_path = role_dir / "index.md"
    skills_path = role_dir / "skills.yml"

    if not index_path.exists():
        failures.append(f"{role_dir}: missing index.md")
        return
    if not skills_path.exists():
        failures.append(f"{role_dir}: missing skills.yml")
        return

    fm, body = parse_frontmatter(index_path.read_text(encoding="utf-8"))
    missing_keys = REQUIRED_INDEX_KEYS - set(fm)
    if missing_keys:
        failures.append(
            f"{index_path}: missing frontmatter keys: {sorted(missing_keys)}"
        )

    first_tasks = count_first_tasks(body)
    if first_tasks < MIN_FIRST_TASKS:
        failures.append(
            f"{index_path}: requires ≥ {MIN_FIRST_TASKS} first tasks, found {first_tasks}"
        )

    prompts_dir = role_dir / "prompts"
    prompts = sorted(prompts_dir.glob("*.md")) if prompts_dir.exists() else []
    if len(prompts) < MIN_PROMPTS_PER_ROLE:
        failures.append(
            f"{role_dir}: requires ≥ {MIN_PROMPTS_PER_ROLE} prompts in prompts/, "
            f"found {len(prompts)}"
        )

    for prompt_path in prompts:
        prompt_fm, _ = parse_frontmatter(prompt_path.read_text(encoding="utf-8"))
        missing = REQUIRED_PROMPT_KEYS - set(prompt_fm)
        if missing:
            failures.append(
                f"{prompt_path}: missing frontmatter keys: {sorted(missing)}"
            )

    try:
        skills_doc = yaml.safe_load(skills_path.read_text(encoding="utf-8")) or {}
    except yaml.YAMLError as exc:
        failures.append(f"{skills_path}: malformed YAML ({exc})")
        return
    skill_entries = skills_doc.get("skills", [])
    if not isinstance(skill_entries, list):
        failures.append(f"{skills_path}: `skills:` must be a list")
        return
    for entry in skill_entries:
        if not isinstance(entry, dict):
            failures.append(f"{skills_path}: skill entry is not a mapping: {entry!r}")
            continue
        skill_id = entry.get("id")
        if not isinstance(skill_id, str):
            failures.append(f"{skills_path}: skill entry missing `id`: {entry!r}")
            continue
        if known_skills and skill_id not in known_skills:
            failures.append(
                f"{skills_path}: skill `{skill_id}` does not resolve to an existing skill"
            )


def lint_plain_language(failures: list[str]) -> None:
    """Scan each role's index.md for the six jargon terms; flag hits.

    Skipped sections: frontmatter, fenced code blocks. Role-experience
    index files are the non-developer surface and must not leak the
    package's internal vocabulary.
    """
    for role_dir in sorted(ROLES_DIR.iterdir()):
        if not role_dir.is_dir() or role_dir.name.startswith(("_", ".")):
            continue
        index_path = role_dir / "index.md"
        if not index_path.exists():
            continue
        _, body = parse_frontmatter(index_path.read_text(encoding="utf-8"))
        # Strip fenced code blocks.
        body_stripped = re.sub(r"```[\s\S]*?```", "", body)
        for term in PLAIN_LANGUAGE_JARGON:
            pattern = re.compile(rf"\b{re.escape(term)}\b", re.IGNORECASE)
            if pattern.search(body_stripped):
                failures.append(
                    f"{index_path}: contains the jargon term `{term}` "
                    f"(see docs/contracts/plain-language-surface.md)"
                )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(prog="lint_role_experiences")
    parser.add_argument(
        "--plain-language",
        action="store_true",
        help="Additionally scan role indices for jargon listed in "
        "docs/contracts/plain-language-surface.md.",
    )
    args = parser.parse_args(argv)

    if not ROLES_DIR.exists():
        print(f"lint_role_experiences: roles dir not found at {ROLES_DIR}")
        return 0

    known_skills = all_skills()
    failures: list[str] = []
    for role_dir in sorted(ROLES_DIR.iterdir()):
        if role_dir.is_dir():
            lint_role(role_dir, known_skills, failures)

    if args.plain_language:
        lint_plain_language(failures)

    if failures:
        for f in failures:
            print(f"❌ {f}")
        print(f"\nlint_role_experiences: {len(failures)} failure(s)")
        return 1
    print("✅ lint_role_experiences: all role experiences pass")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/lint_role_experiences.ts
#!/usr/bin/env tsx
/**
 * Linter for `agents/roles/<slug>/` role experiences.
 *
 * TypeScript twin of `src/scripts/lint_role_experiences.py` (ADR-089,
 * Phase 4 / Wave 4b). The CLI contract is mirrored EXACTLY —
 * `--plain-language` flag, exit codes (0 pass / roles-dir-absent,
 * 1 failures), stdout/stderr split (all output on stdout), byte-identical
 * finding messages (including Python `sorted(set)` list repr), same scan
 * order, same `yaml.safe_load` frontmatter parser, same status↔
 * recruit_session_ref coupling.
 *
 * Asserts the structural floor pinned in docs/contracts/role-experience.md.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { parse as parseYaml } from 'yaml';

const _HERE = fileURLToPath(import.meta.url);

const ROOT = path.resolve(path.dirname(_HERE), '..', '..');
const ROLES_DIR = path.join(ROOT, 'agents', 'roles');
const SKILL_SOURCES = [
    path.join(ROOT, '.agent-src.uncondensed', 'skills'),
    path.join(ROOT, 'dist/agent-src', 'skills'),
];

const REQUIRED_INDEX_KEYS = new Set([
    'role',
    'display_name',
    'tagline',
    'recommended_packs',
    'install_path_hint',
    'recruit_session_ref',
    'status',
]);

const REQUIRED_PROMPT_KEYS = new Set([
    'name',
    'intent',
    'inputs',
    'output_shape',
    'skill_hint',
]);

const VALID_STATUS = new Set(['draft', 'beta-internal', 'beta', 'stable']);
const EXTERNAL_VALIDATED_STATUS = new Set(['beta', 'stable']);

const MIN_FIRST_TASKS = 3;
const MIN_PROMPTS_PER_ROLE = 5;

// re.compile(r"^---\r?\n(.*?)\r?\n---\r?\n", re.DOTALL) anchored at start.
const FRONTMATTER_RE = /^---\r?\n([\s\S]*?)\r?\n---\r?\n/;
const TASK_HEADING_RE = /^## Three first tasks[ \t]*$/m;
const TASK_ITEM_RE = /^[ \t]*\d+\.[ \t]+\*\*[^*]+\*\*/gm;

const PLAIN_LANGUAGE_JARGON = [
    'council',
    'trust level',
    'pack',
    'orchestration',
    'contract',
    'advisory',
];

function parse_frontmatter(text: string): [Record<string, unknown>, string] {
    const m = FRONTMATTER_RE.exec(text);
    if (m === null) {
        return [{}, text];
    }
    let loaded: unknown;
    try {
        loaded = parseYaml(m[1] as string, { version: '1.1' }) ?? {};
    } catch {
        return [{}, text.slice(m[0].length)];
    }
    if (!isPlainObject(loaded)) {
        return [{}, text.slice(m[0].length)];
    }
    return [loaded as Record<string, unknown>, text.slice(m[0].length)];
}

function count_first_tasks(body: string): number {
    TASK_HEADING_RE.lastIndex = 0;
    const headingMatch = TASK_HEADING_RE.exec(body);
    if (headingMatch === null) {
        return 0;
    }
    const after = body.slice(headingMatch.index + headingMatch[0].length);
    const nextRe = /^## /m;
    const nextSection = nextRe.exec(after);
    const section = nextSection ? after.slice(0, nextSection.index) : after;
    TASK_ITEM_RE.lastIndex = 0;
    const matches = section.match(TASK_ITEM_RE);
    return matches ? matches.length : 0;
}

function all_skills(): Set<string> {
    const found = new Set<string>();
    for (const src of SKILL_SOURCES) {
        if (!_exists(src)) {
            continue;
        }
        for (const name of _iterdir(src)) {
            if (_isDir(path.join(src, name))) {
                found.add(name);
            }
        }
    }
    return found;
}

function lint_role(roleDir: string, knownSkills: Set<string>, failures: string[]): void {
    const slug = path.basename(roleDir);
    if (slug.startsWith('_') || slug.startsWith('.')) {
        return;
    }

    const indexPath = path.join(roleDir, 'index.md');
    const skillsPath = path.join(roleDir, 'skills.yml');

    if (!_exists(indexPath)) {
        failures.push(`${roleDir}: missing index.md`);
        return;
    }
    if (!_exists(skillsPath)) {
        failures.push(`${roleDir}: missing skills.yml`);
        return;
    }

    const [fm, body] = parse_frontmatter(fs.readFileSync(indexPath, 'utf-8'));
    const missingKeys = _setDiff(REQUIRED_INDEX_KEYS, new Set(Object.keys(fm)));
    if (missingKeys.size) {
        failures.push(`${indexPath}: missing frontmatter keys: ${_sortedListRepr(missingKeys)}`);
    }

    const status = fm['status'];
    if (typeof status !== 'string' || !VALID_STATUS.has(status)) {
        failures.push(`${indexPath}: status ${_pyRepr(status)} not in ${_sortedListRepr(VALID_STATUS)}`);
    } else if (EXTERNAL_VALIDATED_STATUS.has(status) && !_truthy(fm['recruit_session_ref'])) {
        failures.push(
            `${indexPath}: status '${status}' requires a non-null ` +
                'recruit_session_ref (external-validation gate); use ' +
                "'beta-internal' for the internal-authoring basis",
        );
    }

    const firstTasks = count_first_tasks(body);
    if (firstTasks < MIN_FIRST_TASKS) {
        failures.push(
            `${indexPath}: requires ≥ ${MIN_FIRST_TASKS} first tasks, found ${firstTasks}`,
        );
    }

    const promptsDir = path.join(roleDir, 'prompts');
    const prompts = _exists(promptsDir)
        ? _iterdir(promptsDir)
              .filter((n) => n.endsWith('.md'))
              .map((n) => path.join(promptsDir, n))
              .sort()
        : [];
    if (prompts.length < MIN_PROMPTS_PER_ROLE) {
        failures.push(
            `${roleDir}: requires ≥ ${MIN_PROMPTS_PER_ROLE} prompts in prompts/, ` +
                `found ${prompts.length}`,
        );
    }

    for (const promptPath of prompts) {
        const [promptFm] = parse_frontmatter(fs.readFileSync(promptPath, 'utf-8'));
        const missing = _setDiff(REQUIRED_PROMPT_KEYS, new Set(Object.keys(promptFm)));
        if (missing.size) {
            failures.push(`${promptPath}: missing frontmatter keys: ${_sortedListRepr(missing)}`);
        }
    }

    let skillsDoc: unknown;
    try {
        skillsDoc = parseYaml(fs.readFileSync(skillsPath, 'utf-8'), { version: '1.1' }) ?? {};
    } catch (exc) {
        const msg = exc instanceof Error ? exc.message : String(exc);
        failures.push(`${skillsPath}: malformed YAML (${msg})`);
        return;
    }
    const skillEntries = _get(skillsDoc, 'skills') ?? [];
    if (!Array.isArray(skillEntries)) {
        failures.push(`${skillsPath}: \`skills:\` must be a list`);
        return;
    }
    for (const entry of skillEntries) {
        if (!isPlainObject(entry)) {
            failures.push(`${skillsPath}: skill entry is not a mapping: ${_pyRepr(entry)}`);
            continue;
        }
        const skillId = (entry as Record<string, unknown>)['id'];
        if (typeof skillId !== 'string') {
            failures.push(`${skillsPath}: skill entry missing \`id\`: ${_pyRepr(entry)}`);
            continue;
        }
        if (knownSkills.size && !knownSkills.has(skillId)) {
            failures.push(
                `${skillsPath}: skill \`${skillId}\` does not resolve to an existing skill`,
            );
        }
    }
}

function lint_plain_language(failures: string[]): void {
    for (const name of _iterdir(ROLES_DIR).sort()) {
        const roleDir = path.join(ROLES_DIR, name);
        if (!_isDir(roleDir) || name.startsWith('_') || name.startsWith('.')) {
            continue;
        }
        const indexPath = path.join(roleDir, 'index.md');
        if (!_exists(indexPath)) {
            continue;
        }
        const [, body] = parse_frontmatter(fs.readFileSync(indexPath, 'utf-8'));
        const bodyStripped = body.replace(/```[\s\S]*?```/g, '');
        for (const term of PLAIN_LANGUAGE_JARGON) {
            const pattern = new RegExp(`\\b${_reEscape(term)}\\b`, 'i');
            if (pattern.test(bodyStripped)) {
                failures.push(
                    `${indexPath}: contains the jargon term \`${term}\` ` +
                        '(see docs/contracts/plain-language-surface.md)',
                );
            }
        }
    }
}

interface Args {
    plainLanguage: boolean;
}

function parse_args(argv: readonly string[]): Args {
    let plainLanguage = false;
    for (const arg of argv) {
        if (arg === '--plain-language') {
            plainLanguage = true;
        } else if (arg === '-h' || arg === '--help') {
            process.stdout.write('usage: lint_role_experiences [-h] [--plain-language]\n');
            process.exit(0);
        } else {
            process.stderr.write(`lint_role_experiences: error: unrecognized arguments: ${arg}\n`);
            process.exit(2);
        }
    }
    return { plainLanguage };
}

function main(argv?: readonly string[]): number {
    const args = parse_args(argv ?? process.argv.slice(2));

    if (!_exists(ROLES_DIR)) {
        process.stdout.write(`lint_role_experiences: roles dir not found at ${ROLES_DIR}\n`);
        return 0;
    }

    const knownSkills = all_skills();
    const failures: string[] = [];
    for (const name of _iterdir(ROLES_DIR).sort()) {
        const roleDir = path.join(ROLES_DIR, name);
        if (_isDir(roleDir)) {
            lint_role(roleDir, knownSkills, failures);
        }
    }

    if (args.plainLanguage) {
        lint_plain_language(failures);
    }

    if (failures.length) {
        for (const f of failures) {
            process.stdout.write(`❌ ${f}\n`);
        }
        process.stdout.write(`\nlint_role_experiences: ${failures.length} failure(s)\n`);
        return 1;
    }
    process.stdout.write('✅ lint_role_experiences: all role experiences pass\n');
    return 0;
}

// --- helpers --------------------------------------------------------------

function _exists(p: string): boolean {
    try {
        fs.statSync(p);
        return true;
    } catch {
        return false;
    }
}

function _isDir(p: string): boolean {
    try {
        return fs.statSync(p).isDirectory();
    } catch {
        return false;
    }
}

function _iterdir(p: string): string[] {
    try {
        return fs.readdirSync(p);
    } catch {
        return [];
    }
}

function isPlainObject(v: unknown): v is Record<string, unknown> {
    return typeof v === 'object' && v !== null && !Array.isArray(v);
}

function _get(obj: unknown, key: string): unknown {
    return isPlainObject(obj) ? (obj as Record<string, unknown>)[key] : undefined;
}

/** Python truthiness for `fm.get("recruit_session_ref")` — null/""/0/false → false. */
function _truthy(v: unknown): boolean {
    if (v === null || v === undefined || v === false) {
        return false;
    }
    if (v === '') {
        return false;
    }
    if (v === 0) {
        return false;
    }
    if (Array.isArray(v)) {
        return v.length > 0;
    }
    if (isPlainObject(v)) {
        return Object.keys(v).length > 0;
    }
    return true;
}

function _setDiff(a: Set<string>, b: Set<string>): Set<string> {
    const out = new Set<string>();
    for (const x of a) {
        if (!b.has(x)) {
            out.add(x);
        }
    }
    return out;
}

function _reEscape(s: string): string {
    return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/** Mirror Python `sorted(set)` repr as a list literal of single-quoted strings. */
function _sortedListRepr(s: Set<string>): string {
    return '[' + [...s].sort().map((x) => `'${x}'`).join(', ') + ']';
}

/** Python repr() for a scalar value embedded in a finding (status {!r}). */
function _pyRepr(value: unknown): string {
    if (value === null || value === undefined) {
        return 'None';
    }
    if (value === true) {
        return 'True';
    }
    if (value === false) {
        return 'False';
    }
    if (typeof value === 'number') {
        return String(value);
    }
    if (typeof value === 'string') {
        return `'${value.replace(/\\/g, '\\\\').replace(/'/g, "\\'")}'`;
    }
    if (Array.isArray(value)) {
        return '[' + value.map((v) => _pyRepr(v)).join(', ') + ']';
    }
    if (isPlainObject(value)) {
        const parts = Object.entries(value as Record<string, unknown>).map(
            ([k, v]) => `${_pyRepr(k)}: ${_pyRepr(v)}`,
        );
        return '{' + parts.join(', ') + '}';
    }
    return String(value);
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main());
}

export {
    ROOT,
    ROLES_DIR,
    REQUIRED_INDEX_KEYS,
    REQUIRED_PROMPT_KEYS,
    VALID_STATUS,
    EXTERNAL_VALIDATED_STATUS,
    parse_frontmatter,
    count_first_tasks,
    all_skills,
    lint_role,
    lint_plain_language,
    parse_args,
    main,
};
